### PR TITLE
Fix require in section

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

File: doc/changelog/07-commands-and-options/11972-fix-require-in-section.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- **Changed:**
+  Using :cmd:`Require` inside a section is not deprecated anymore.
```
Comment (Zimmi48) [inline doc/changelog/07-commands-and-options/11972-fix-require-in-section.rst]: What about:
```suggestion
- **Fixed:**
  Using :cmd:`Require` inside a section has been fixed, and thus is not deprecated anymore
```

**Comment by ejgallego** [comment]
Not sure this is safe w.r.t. plugins; if the `Require` is undone on section closing; the summary may still contain the registered state for that module, see `warn_summary_out_of_scope`

**Comment by ppedrot** [comment]
> AFAICT there is no point in deprecating it.

People who tried to fix Libobject might be disagreeing. @maximedenes ?

**Comment by SkySkimmer** [comment]
I removed the undeprecation so this is just the fix

**Comment by SkySkimmer** [comment]
ping @coq/kernel-maintainers

**Review by ppedrot — APPROVED** [review]
I can't swear this is fully correct since there could be fancy interactions with arbitrary code execution, but it's probably correct enough. I would definitely recommend removing requires in sections at some point in the near future, because it's begging the stick to get beaten.

File: kernel/safe_typing.ml [hunk 2]
```diff
@@ -1351,14 +1286,93 @@ let import lib cst vodigest senv =
     Modops.add_linked_module mb linkinfo env
   in
   let env = Environ.add_native_symbols lib.comp_name lib.comp_natsymbs env in
+  let sections =
+    Option.map (Section.map_custom (fun custom ->
+        {custom with rev_reimport = (lib,cst,vodigest) :: custom.rev_reimport}))
```
Comment (ppedrot) [inline kernel/safe_typing.ml:1291]: For clarity, shouldn't we share the declaration of universe constraints with the mechanism already there for sections? That is, when replaying requires, we shouldn't push any constraints. If so, the code below needs to be adapted by switching replaying of global constraints and requires.

Comment (SkySkimmer) [reply kernel/safe_typing.ml:1291]: I don't see how that's clearer.

Comment (ppedrot) [reply kernel/safe_typing.ml:1291]: Well, at least we don't have *two* mechanisms registering monomorphic constraints. This is definitely asking for trouble, even if currently correct.

Now that I think of it, the current PR might even be unsound. What happens if you require inside a polymorphic section a carefully crafted module that refers to local universe variables?

**Review by ppedrot — CHANGES_REQUESTED** [review]
Please factor out the addition of global constraints with the already existing mechanism. Having a common global state split in two different places is a recipe for disaster.

**Comment by SkySkimmer** [comment]
>Please factor out the addition of global constraints with the already existing mechanism. Having a common global state split in two different places is a recipe for disaster.

I think it's better for sections to treat the arguments of `import` opaquely, not merging with the section level stuff.

>Now that I think of it, the current PR might even be unsound. What happens if you require inside a polymorphic section a carefully crafted module that refers to local universe variables?

Maybe we should enforce that added universes have the expected dirpath?

**Comment by ppedrot** [comment]
> I think it's better for sections to treat the arguments of import opaquely, not merging with the section level stuff.

Global universes are a different sort of animal, they're not really part of the "library" per se. We keep storing them haphazardly in different places, but this is really bad engineering-wise since we keep forgetting that there are there. I insist, please decompose `import` into adding constraints and then loading the relevant code.

**Comment by SkySkimmer** [comment]
I would need to get the `mod_constraints` too, that's too messy for no gain.

**Comment by ppedrot** [comment]
Maybe there is some misunderstanding, my request is really minimal. I can push a minor patch atop of this one to show you.

**Comment by SkySkimmer** [comment]
OK

Gaëtan Gilbert

**Comment by ppedrot** [comment]
@SkySkimmer done. We should really factorize the code adding global constraints in Safe_typing, we're still doing it recklessly.

File: kernel/safe_typing.ml [hunk 3]
```diff
@@ -1332,33 +1267,115 @@ let export ?except ~output_native_objects senv dir =
...
     modresolver = Mod_subst.add_delta_resolver mb.mod_delta senv.modresolver;
     required = DPmap.add lib.comp_name vodigest senv.required;
     loads = (mp,mb)::senv.loads;
+    sections;
   }
 
+(* cst are the constraints that were computed by the vi2vo step and hence are
+ * not part of the mb.mod_constraints field (but morally should be) *)
+let import lib cst vodigest senv =
+  let senv = push_context_set ~strict:true lib.comp_mod.mod_constraints senv in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml:1302]: Using Safe_typing.push_context_set means we add the dependency's universes to the current module's universes, that seems wrong to me.

Comment (ppedrot) [reply kernel/safe_typing.ml:1302]: You're right indeed.

Comment (ppedrot) [reply kernel/safe_typing.ml:1302]: That said, if I understand the code correctly, nested modules transitively contain the universe constraints of their subcomponents. Really, this global state shouldn't be stored this way.

**Comment by ppedrot** [comment]
@SkySkimmer OK, fine, your PR was probably mergeable in its original state. Still, I am utterly in favour of removing Require-in-section instead of undeprecating it.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**